### PR TITLE
Add oom retry handling for createGatherer in gpu hash joins

### PR DESCRIPTION
**Pull request by jbrennan333**
This adds oom retry handling for gpu hash joins.  It is part of the work described in https://github.com/NVIDIA/spark-rapids/issues/7255.

This uses the RmmRapidsRetryIterator framework to add retries for createGatherer`.

For `createGatherer`, some refactoring was done to ensure the buffers are spillable before we go into a retry loop.  There are two paths for the stream side - the original `LazySpillableColumnar` batch from the stream-side iterator, and the pending list (from prior splits), which was a `SpillableColumnarBatch`.  For this pr, I chose to make all of these `LazySpillableColumnarBatches`, because that is what we ultimately hand off to the gatherer.  

I am putting this up as a draft because I have not added unit tests yet, and there is still a bug in this code that causes us to leak column vectors in a few of the broadcast-nested-loop-join integration tests (and when running nds).
I also need to do a performance impact check.

### CODE REVIEW

**Comment by jbrennan333** [comment]
These are the integration tests that leak column vectors with this code:
Current integration test failures:
```
../../src/main/python/join_test.py::test_right_broadcast_nested_loop_join_with_ast_condition[Inner-Boolean-1g][IGNORE_ORDER({'local': True})] 23/03/20 15:13:57 ERROR ColumnVector: A DEVICE COLUMN VECTOR WAS LEAKED (ID: 6506 7fae5c361f80)

../../src/main/python/join_test.py::test_right_broadcast_nested_loop_join_with_ast_condition[LeftSemi-Boolean-1g][IGNORE_ORDER({'local': True})] 23/03/20 15:14:03 ERROR ColumnVector: A DEVICE COLUMN VECTOR WAS LEAKED (ID: 17027 7fae6814f480)

../../src/main/python/join_test.py::test_right_broadcast_nested_loop_join_with_ast_condition[Cross-Integer-100][IGNORE_ORDER({'lo
cal': True})] 23/03/20 15:14:17 WARN GpuOverrides: 
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala [hunk 1]
```diff
@@ -198,22 +199,26 @@ class ConditionalNestedLoopJoinIterator(
...
+      withRetryNoSplit {
+        withResource(GpuColumnVector.from(builtBatch.getBatch)) { builtTable =>
+          withResource(GpuColumnVector.from(streamBatch.getBatch)) { streamTable =>
+            val builtSpillOnly = LazySpillableColumnarBatch.spillOnly(builtBatch)
+            val (leftTable, leftBatch, rightTable, rightBatch) = buildSide match {
+              case GpuBuildLeft => (builtTable, builtSpillOnly, streamTable, streamBatch)
+              case GpuBuildRight => (streamTable, streamBatch, builtTable, builtSpillOnly)
+            }
+            val maps = computeGatherMaps(leftTable, rightTable, numJoinRows)
+            makeGatherer(maps, leftBatch, rightBatch, joinType)
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala:220]: There is a problem here for retries.  makeGatherer can close one of the batches in some cases.  If that happens to the streamBatch, it will not work on a retry.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala [hunk 2]
```diff
@@ -287,11 +288,14 @@ abstract class BaseHashJoinIterator(
...
-      cb: ColumnarBatch,
+      streamBatch: LazySpillableColumnarBatch,
       numJoinRows: Option[Long]): Option[JoinGatherer] = {
     try {
-      withResource(GpuProjectExec.project(built.getBatch, boundBuiltKeys)) { builtKeys =>
-        joinGatherer(builtKeys, built, cb)
+      streamBatch.allowSpilling()
+      withRetryNoSplit {
+        withResource(GpuProjectExec.project(built.getBatch, boundBuiltKeys)) { builtKeys =>
+          joinGatherer(builtKeys, built, streamBatch)
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:297]: This call could close one of the batches.  If it closes the built batch we are ok, but if it closes the streamBatch, that will break on retries.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala [hunk 3]
```diff
@@ -198,22 +199,26 @@ class ConditionalNestedLoopJoinIterator(
...
-          val builtSpillOnly = LazySpillableColumnarBatch.spillOnly(builtBatch)
-          val (leftTable, leftBatch, rightTable, rightBatch) = buildSide match {
-            case GpuBuildLeft => (builtTable, builtSpillOnly, streamTable, streamBatch)
-            case GpuBuildRight => (streamTable, streamBatch, builtTable, builtSpillOnly)
+    closeOnExcept(streamBatch) { streamBatch =>
+      streamBatch.allowSpilling()
+      withRetryNoSplit {
+        withResource(GpuColumnVector.from(builtBatch.getBatch)) { builtTable =>
+          withResource(GpuColumnVector.from(streamBatch.getBatch)) { streamTable =>
+            val builtSpillOnly = LazySpillableColumnarBatch.spillOnly(builtBatch)
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala:214]: Can move this builtSpillOnly assignment before the retry loop.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala [hunk 4]
```diff
@@ -256,10 +250,30 @@ abstract class SplittableJoinIterator(
...
+        lazyCb
+      }
+    }
+  }
+
   /**
    * Split a stream-side input batch, making all splits spillable, and replacing this batch with
    * the splits in the stream-side input
-   * @param cb stream-side input batch to split
+   * @param cb stream-side input batch to split, which will be closed
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala:276]: Need to revert this change to the comment.

**Comment by jbrennan333** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala [hunk 5]
```diff
@@ -200,48 +201,41 @@ abstract class SplittableJoinIterator(
...
-            lazyBatch.releaseBatch()
-          }
-        }
-        batch
+        stream.next()
       }
       opTime.ns {
-        withResource(cb) { cb =>
-          val numJoinRows = computeNumJoinRows(cb)
+        withResource(scb) { scb =>
```
[the remark this answers is not in the document]
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala:210]: Thanks for testing it!  I am trying to repro this.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala [hunk 6]
```diff
@@ -256,10 +250,30 @@ abstract class SplittableJoinIterator(
...
+      cb: ColumnarBatch,
+      numBatches: Int): Seq[LazySpillableColumnarBatch] = {
+    val batchSize = cb.numRows() / numBatches
+    val splits = withResource(GpuColumnVector.from(cb)) { tab =>
+      val splitIndexes = (1 until numBatches).map(num => num * batchSize)
+      tab.contiguousSplit(splitIndexes: _*)
+    }
+    withResource(splits) { splits =>
+      val schema = GpuColumnVector.extractTypes(cb)
+      val tables = splits.map(_.getTable)
```
[the remark this answers is not in the document]
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala:263]: I think this is covered by the withResource(splits).  These tables are closed when we close the ContiguousTables.

**Comment by jbrennan333** [comment]
I am going to split this draft PR into two parts.  I will use this one for the `createGatherer` and related changes.
I have filed #7930 to address the `nextCbFromGatherer` changes.  I am going to focus on #7930 first, because it is a smaller change and I think has more impact (it covers the `ai.rapids.cudf.Table.gather` case and the `getRowsInNextBatch` case.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala]: There was a bug here - thanks!

**Comment by jbrennan333** [comment]
Testing I have done so far includes:
- I have verified correct results for a power run on my desktop at scale 100 with forced retries in both createGatherer implementations.
- I have run the full join integration test suite with forced retry ooms in both createGatherer implementation, and there were no failures.
- I ran a performance check on spark2a and there was a small change:
```
Name = benchmark
Means = 487719.9972629547, 489457.8136444092
Time diff = -1737.8163814544678
Speedup = 0.9964495073262494
```

**Comment by jbrennan333** [comment]
build

**Review by revans2 — DISMISSED** [review]
All of this looks great. I just want to run a few tests locally myself too

**Comment by revans2** [comment]
I am seeing a few failures related to close being called too many times. Specifically when running NDS query 40 with only 4 GiB of memory. I saw 9 failures related to this and all of them has the exact same stack trace. If you need or want more help debugging this please let me know.

```
23/03/30 16:40:43 WARN TaskSetManager: Lost task 8.0 in stage 900.0 (TID 3628) (10.28.9.123 executor 0): java.lang.IllegalStateException: Close called too many times ColumnVector{rows=22865493, type=INT32, nullCount=Optional.empty, offHea
p=(ID: 957712 0)}
        at ai.rapids.cudf.ColumnVector.close(ColumnVector.java:224)
        at com.nvidia.spark.rapids.GpuColumnVector.close(GpuColumnVector.java:1124)
        at org.apache.spark.sql.vectorized.ColumnarBatch.close(ColumnarBatch.java:44)
        at com.nvidia.spark.rapids.LazySpillableColumnarBatchImpl.$anonfun$close$1(JoinGatherer.scala:311)
        at com.nvidia.spark.rapids.LazySpillableColumnarBatchImpl.$anonfun$close$1$adapted(JoinGatherer.scala:311)
        at scala.Option.foreach(Option.scala:407)
        at com.nvidia.spark.rapids.LazySpillableColumnarBatchImpl.close(JoinGatherer.scala:311)
        at com.nvidia.spark.rapids.RapidsPluginImplicits$AutoCloseableColumn.safeClose(implicits.scala:56)
        at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:30)
        at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
        at com.nvidia.spark.rapids.AbstractGpuJoinIterator.withResource(AbstractGpuJoinIterator.scala:54)
        at com.nvidia.spark.rapids.SplittableJoinIterator.$anonfun$setupNextGatherer$1(AbstractGpuJoinIterator.scala:214)
        at com.nvidia.spark.rapids.GpuMetric.ns(GpuExec.scala:131)
        at com.nvidia.spark.rapids.SplittableJoinIterator.setupNextGatherer(AbstractGpuJoinIterator.scala:214)
        at com.nvidia.spark.rapids.AbstractGpuJoinIterator.hasNext(AbstractGpuJoinIterator.scala:96)
        at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
        at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
```

**Comment by jbrennan333** [comment]
@revans2 I have merged up to include your latest changes, and added a possible fix for the double-close issue.  Can you please try again with this version when you get a chance?  I'm still having trouble reproducing locally.

**Comment by jbrennan333** [comment]
build

**Comment by jbrennan333** [comment]
build

**Comment by jbrennan333** [comment]
@revans2 I was finally able to reproduce the inc-after-close,  double-close by forcing an oom during the checkpoint of the stream batch.

Moving the `checkpoint` calls (which call `allowSpilling` for `LazySpillableColumnarBatch`) outside of the try block was part of the fix - I no longer get the inc-after-close, but I was still hitting the double-close.

The other part of the fix is in `LazySpillableColumnarBatchImpl.allowSpilling()`.  It needs to ensure it clears the cached value if we throw while creating the `SpillableColumnarBatch`.  I made this change for `LazySpillableGatherMapImpl.allowSpilling()` as well.

I think this might fix the double-close reported in #7581 

Thanks again for testing this and providing logs to help me debug it.

**Comment by jbrennan333** [comment]
build

**Comment by jbrennan333** [comment]
build

**Review by revans2 — APPROVED** [review]

**Review by abellina — APPROVED** [review]

**Comment by revans2** [comment]
Looks good. I am seeing a few leaked columns when I do my testing with really low memory, but none of them are even close to the join code, so I think we are good. I'll try to track the others down and file something separately for them
